consensus/misc/eip1559/eip1559_optimism.go

```diff
@@ -0,0 +1,160 @@
+	}
+	// leave version byte 0
+	binary.BigEndian.PutUint32(r[1:5], uint32(denom))
+	binary.BigEndian.PutUint32(r[5:], uint32(elasticity))
+	return r
+}
+
+// ValidateHolocene1559Params checks if the encoded parameters are valid according to the Holocene
+// upgrade.
+func ValidateHolocene1559Params(params []byte) error {
+	if len(params) != 8 {
+		return fmt.Errorf("holocene eip-1559 params should be 8 bytes, got %d", len(params))
+	}
+	d, e := DecodeHolocene1559Params(params)
+	if e != 0 && d == 0 {
+		return errors.New("holocene params cannot have a 0 denominator unless elasticity is also 0")
+	}
+	return nil
+}
+
+// ValidateHoloceneExtraData checks if the header extraData is valid according to the Holocene
+// upgrade.
+func ValidateHoloceneExtraData(extra []byte) error {
+	if len(extra) != 9 {
+		return fmt.Errorf("holocene extraData should be 9 bytes, got %d", len(extra))
+	}
+	if extra[0] != 0 {
+		return fmt.Errorf("holocene extraData should have 0 version byte, got %d", extra[0])
+	}
+	return ValidateHolocene1559Params(extra[1:])
+}
+
+// DecodeJovianExtraData decodes the extraData parameters from the encoded form defined here:
+// https://specs.optimism.io/protocol/jovian/exec-engine.html
+//
+// Returns 0,0,0 if the format is invalid, though ValidateMinBaseFeeExtraData should be used instead of this function for
+// validity checking.
+func DecodeJovianExtraData(extra []byte) (uint64, uint64, uint64) {
+	// Best effort to decode the extraData for every block in the chain's history,
+	// including blocks before the minimum base fee feature was enabled.
+	if len(extra) == 9 {
+		// This is Holocene extraData
+		denominator, elasticity := DecodeHolocene1559Params(extra[1:9])
+		return denominator, elasticity, 0
+	} else if len(extra) == 17 {
+		// Decode extraData when the minimum base fee fork is enabled
+		denominator, elasticity := DecodeHolocene1559Params(extra[1:9])
+		minBaseFee := binary.BigEndian.Uint64(extra[9:])
+		return denominator, elasticity, minBaseFee
+	}
+	return 0, 0, 0
+}
+
+// EncodeJovianExtraData encodes the EIP-1559 and minBaseFee parameters into the header 'ExtraData' format.
+// Will panic if EIP-1559 parameters are outside uint32 range.
+func EncodeJovianExtraData(denom, elasticity, minBaseFee uint64) []byte {
+	r := make([]byte, 17)
+	if denom > gomath.MaxUint32 || elasticity > gomath.MaxUint32 {
+		panic("eip-1559 parameters out of uint32 range")
+	}
+	r[0] = 1
+	binary.BigEndian.PutUint32(r[1:5], uint32(denom))
+	binary.BigEndian.PutUint32(r[5:9], uint32(elasticity))
+	binary.BigEndian.PutUint64(r[9:], minBaseFee)
+	return r
+}
+
+// ValidateJovianExtraData checks if the header extraData is valid according to the minimum base fee feature.
+func ValidateJovianExtraData(extra []byte) error {
+	if len(extra) != 17 {
+		return fmt.Errorf("minBaseFee extraData should be 17 bytes, got %d", len(extra))
+	}
+	if extra[0] != 1 {
+		return fmt.Errorf("minBaseFee version should be 1, got %d", extra[0])
```

Comment:
```suggestion
		return fmt.Errorf("Jovian extraData version byte should be 1, got %d", extra[0])
```
probably also worth cleaning up the Holocene error message to be similar, the "should have 0 version byte" sounds ambiguous.

Reply:
[109385a](https://github.com/ethereum-optimism/op-geth/pull/666/commits/109385a90f47161738e3e1e9e42dbe0c953a88bc)